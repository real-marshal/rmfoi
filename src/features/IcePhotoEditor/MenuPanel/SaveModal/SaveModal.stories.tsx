import type { Meta, Story } from '@storybook/react'
import { SaveModal, SaveModalProps } from './SaveModal'

export default {
  title: 'Common/SaveModal',
  component: SaveModal,
} as Meta

const Template: Story<SaveModalProps> = (args) => <SaveModal {...args} />

export const Basic = Template.bind({})
Basic.args = {}
