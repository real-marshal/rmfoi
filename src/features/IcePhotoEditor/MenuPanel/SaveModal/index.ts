export { SaveModal } from './SaveModal'
export type { SaveModalProps } from './SaveModal'
