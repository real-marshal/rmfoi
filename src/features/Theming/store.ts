import type { RootState } from '@/app/store'
import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import { SystemThemeValue, SYSTEM_THEME_FALLBACK, ThemeValue } from './constants'
import storage from 'redux-persist/lib/storage'
import { PersistConfig, persistReducer } from 'redux-persist'

const NAME = 'theme' as const
interface ThemeState {
  currentTheme: ThemeValue
  systemTheme: SystemThemeValue
}

const initialState: ThemeState = {
  currentTheme: ThemeValue.SYSTEM,
  systemTheme: SYSTEM_THEME_FALLBACK,
}

const persistConfig: PersistConfig<ThemeState> = {
  key: NAME,
  storage,
} as const

/*

REDUCERS

*/

const { actions, reducer } = createSlice({
  name: NAME,
  initialState,
  reducers: {
    setCurrentTheme: (state, action: PayloadAction<ThemeValue>) => {
      state.currentTheme = action.payload
    },
    setSystemTheme: (state, action: PayloadAction<SystemThemeValue>) => {
      state.systemTheme = action.payload
    },
  },
})

export { actions, reducer }
export const themeReducer = persistReducer(persistConfig, reducer)

/*

SELECTORS

*/

export const selectCurrentTheme = (state: RootState) => state.theme.currentTheme

export const selectActualTheme = (state: RootState) =>
  state.theme.currentTheme === ThemeValue.SYSTEM
    ? state.theme.systemTheme
    : state.theme.currentTheme
