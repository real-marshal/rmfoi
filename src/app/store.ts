import { themeReducer } from '@/features/Theming'
import {
  AnyAction,
  combineReducers,
  configureStore,
  Dispatch,
  Reducer,
  ThunkAction,
} from '@reduxjs/toolkit'
import type { ThunkMiddlewareFor } from '@reduxjs/toolkit/dist/getDefaultMiddleware'
import type { ExtractDispatchExtensions } from '@reduxjs/toolkit/dist/tsHelpers'
import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux'
import { FLUSH, PAUSE, PERSIST, persistStore, PURGE, REGISTER, REHYDRATE } from 'redux-persist'
import { get, set } from '../utils'
import { produce } from 'immer'

const nonSerializableFields = [
  {
    key: 'icePhotoEditor.image',
    replacement: '<ImageData>',
  },
  {
    key: 'icePhotoEditor.currentImage',
    replacement: '<ImageData>',
  },
]

const staticReducers = {
  theme: themeReducer,
}

const store = configureStore({
  reducer: staticReducers,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
        ignoredPaths: nonSerializableFields.map(({ key }) => key),
      },
      immutableCheck: {
        ignoredPaths: nonSerializableFields.map(({ key }) => key),
      },
    }),
  devTools: {
    stateSanitizer: (state) => {
      const sanitize = produce<typeof state>((draft) => {
        nonSerializableFields.forEach(({ key, replacement }) => {
          if (get(draft, key)) {
            set(draft as object, key, replacement)
          }
        })
      })

      return sanitize(state)
    },
  },
})

const persistor = persistStore(store)

const asyncReducers: Record<string, Reducer> = {}

store.injectReducer = (key: string, asyncReducer: Reducer) => {
  asyncReducers[key] = asyncReducer

  const rootReducer = combineReducers({
    ...staticReducers,
    ...asyncReducers,
  })

  store.replaceReducer(rootReducer)
}

export default store

export { persistor }

// Types for global synchronous store
export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch
// eslint-disable-next-line @typescript-eslint/no-invalid-void-type
export type AppThunk<ReturnType = void> = ThunkAction<ReturnType, RootState, unknown, AnyAction>

type Middlewares<S = RootState> = [ThunkMiddlewareFor<S>]

// Type factories for asynchronously loaded slices
export type MakeCurrentState<State> = RootState & State
export type MakeCurrentDispatch<State> = ExtractDispatchExtensions<
  Middlewares<MakeCurrentState<State>>
> &
  Dispatch
export type MakeCurrentThunk<State, ReturnType> = ThunkAction<
  ReturnType,
  MakeCurrentState<State>,
  unknown,
  AnyAction
>

export const useAppDispatch: () => AppDispatch = useDispatch
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector
