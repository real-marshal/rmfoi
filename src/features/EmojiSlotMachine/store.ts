import store, { MakeCurrentDispatch, MakeCurrentState, MakeCurrentThunk } from '@/app/store'
import { createSelector, createSlice, PayloadAction } from '@reduxjs/toolkit'
import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux'
import { Emoji, emojis, EmojiSlotMachineMode, modeData } from './constants'
import { getShuffledEmojiArrays, getWonEmojiIndexes, EmojiSwapData, tryToWin } from './utils'

const NAME = 'emojiSlotMachine' as const

interface EmojiSlotMachineState {
  mode: EmojiSlotMachineMode
  columns: number
  emojiArrays: Emoji[][]
  chance: number
  chanceIncrement: number
  minIndexToWin: number
  maxIndexToWin: number
  stats: {
    pulls: number
    wonEmojis: Partial<Record<Emoji, number>>
  }
}

// This slice will be injected asyncronously which is why we need separate types here
interface SliceState {
  [NAME]: EmojiSlotMachineState
}
export type CurrentRootState = MakeCurrentState<SliceState>
export type CurrentDispatch = MakeCurrentDispatch<SliceState>
// eslint-disable-next-line @typescript-eslint/no-invalid-void-type
export type CurrentThunk<ReturnType = void> = MakeCurrentThunk<SliceState, ReturnType>

export const useCurrentDispatch: () => CurrentDispatch = useDispatch
export const useCurrentSelector: TypedUseSelectorHook<CurrentRootState> = useSelector

const initialState: EmojiSlotMachineState = {
  mode: EmojiSlotMachineMode.STANDARD,
  columns: 3,
  emojiArrays: getShuffledEmojiArrays(3),
  minIndexToWin: 0,
  maxIndexToWin: emojis.length - 1,
  stats: {
    pulls: 0,
    wonEmojis: {},
  },
  ...modeData[EmojiSlotMachineMode.STANDARD],
}

/*

REDUCERS

*/

const { actions, reducer } = createSlice({
  name: NAME,
  initialState,
  reducers: {
    initialize: (
      state,
      action: PayloadAction<
        Partial<Pick<EmojiSlotMachineState, 'columns' | 'mode' | 'minIndexToWin' | 'maxIndexToWin'>>
      >
    ) => {
      Object.assign(state, action.payload)
    },
    setColumns: (state, action: PayloadAction<number>) => {
      state.columns = action.payload
    },
    setEmojiArrays: (state, action: PayloadAction<EmojiSlotMachineState['emojiArrays']>) => {
      state.emojiArrays = action.payload
    },
    win: (state, action: PayloadAction<Emoji>) => {
      state.stats.pulls++

      Object.assign(state, modeData[state.mode])

      const wonEmoji = state.stats.wonEmojis[action.payload]
      state.stats.wonEmojis[action.payload] = wonEmoji ? wonEmoji + 1 : 1
    },
    lose: (state) => {
      state.stats.pulls++
      state.chance += state.chanceIncrement
    },
    swapEmojis: (state, action: PayloadAction<EmojiSwapData[]>) => {
      const { emojiArrays } = state

      action.payload.forEach(
        ({ arrayIndex, emojiInd1, emojiInd2 }) =>
          ([emojiArrays[arrayIndex]![emojiInd1]!, emojiArrays[arrayIndex]![emojiInd2]!] = [
            emojiArrays[arrayIndex]![emojiInd2]!,
            emojiArrays[arrayIndex]![emojiInd1]!,
          ])
      )
    },
    updateMode: (state, action: PayloadAction<EmojiSlotMachineMode>) => {
      state.mode = action.payload
      Object.assign(state, modeData[state.mode])
    },
  },
})

export { actions }

store.injectReducer(NAME, reducer)

/*

THUNKS

*/

// shuffleEmojiArrays has impure RNG logic so it shouldn't be in the reducer
export const shuffleEmojiArrays = (): CurrentThunk => (dispatch, getState) => {
  const { columns } = getState()[NAME]
  const shuffledEmojis = getShuffledEmojiArrays(columns)

  dispatch(actions.setEmojiArrays(shuffledEmojis))
}

export const updateColumns =
  (columns: number): CurrentThunk =>
  (dispatch) => {
    dispatch(actions.setColumns(columns))
    dispatch(shuffleEmojiArrays())
  }

export const evaluateNewPull =
  (): CurrentThunk<{ wonEmojiIndexes: number[]; wonEmoji: Emoji } | undefined> =>
  (dispatch, getState) => {
    dispatch(shuffleEmojiArrays())

    const { chance, minIndexToWin, maxIndexToWin, emojiArrays } = getState()[NAME]
    const wonEmoji = tryToWin({ chance, minIndexToWin, maxIndexToWin })

    if (!wonEmoji) {
      dispatch(actions.lose())
      return
    }

    dispatch(actions.win(wonEmoji))

    const { wonEmojiIndexes, emojisToSwap } = getWonEmojiIndexes({
      emojiArrays,
      wonEmoji,
      minIndexToWin,
      maxIndexToWin,
    })
    dispatch(actions.swapEmojis(emojisToSwap))

    return { wonEmojiIndexes, wonEmoji }
  }

/*

SELECTORS

*/

export const selectEmojiArrays = (state: CurrentRootState) => state[NAME].emojiArrays

export const selectMode = (state: CurrentRootState) => state[NAME].mode

export const selectChance = (state: CurrentRootState) => state[NAME].chance

export const selectChanceIncrement = (state: CurrentRootState) => state[NAME].chanceIncrement

export const selectStats = (state: CurrentRootState) => state[NAME].stats

export const selectGeneralStats = createSelector(
  selectStats,
  selectChance,
  selectChanceIncrement,
  ({ pulls, wonEmojis }, chance, chanceIncrement) => {
    const wins = Object.values(wonEmojis).reduce((result, wins) => result + wins, 0)

    return {
      pulls,
      wins,
      loses: pulls - wins,
      WR: pulls ? wins / pulls : 0,
      chance,
      chanceIncrement,
    }
  }
)

export const selectdEmojiStats = createSelector(selectStats, ({ wonEmojis }) => {
  const wonEmojisNumber = Object.keys(wonEmojis).length

  return {
    wonEmojisNumber,
    leftEmojis: emojis.length - wonEmojisNumber,
    wonEmojis,
  }
})
