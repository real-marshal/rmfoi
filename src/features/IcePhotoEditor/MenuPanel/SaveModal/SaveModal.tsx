import { Button, Heading, Modal } from '@/components'
import { dt } from '@/features/Theming'
import styled from '@emotion/styled'
import { MouseEventHandler, useEffect, useState } from 'react'
import { selectFilename, useCurrentSelector } from '../../store'

interface SaveModalProps {
  isOpen: boolean
  onSave: (filename: string) => void
  onCancel: MouseEventHandler
}

const ModalContainer = styled.div`
  display: flex;
  flex-direction: column;
  gap: ${dt.gap.md};
  padding: ${dt.padding.md};
`

const FilenameContainer = styled.div`
  display: flex;
  gap: ${dt.gap.sm};
  align-items: center;
`
const FilenameLabel = styled.label``
const FilenameInput = styled.input``

const ButtonsContainer = styled.div`
  display: flex;
  justify-content: space-evenly;
`

const SaveModal = ({ isOpen, onSave, onCancel }: SaveModalProps) => {
  const originalFilename = useCurrentSelector(selectFilename)

  const [filename, setFilename] = useState(originalFilename ?? '')

  useEffect(() => {
    const filenameWithoutExtension = originalFilename?.split('.').slice(0, -1).join('')

    setFilename(filenameWithoutExtension ?? '')
  }, [originalFilename])

  return (
    <Modal isOpen={isOpen}>
      <ModalContainer>
        <Heading type='h1'>Save the image</Heading>
        <FilenameContainer>
          <FilenameLabel>Filename</FilenameLabel>
          <FilenameInput value={filename} onChange={(e) => setFilename(e.target.value)} />
        </FilenameContainer>
        <ButtonsContainer>
          <Button onClick={() => onSave(filename)}>Save</Button>
          <Button secondary onClick={onCancel}>
            Cancel
          </Button>
        </ButtonsContainer>
      </ModalContainer>
    </Modal>
  )
}

export { SaveModal }
export type { SaveModalProps }
