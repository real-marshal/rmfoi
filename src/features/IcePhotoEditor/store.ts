import store, { MakeCurrentDispatch, MakeCurrentState, MakeCurrentThunk } from '@/app/store'
import { createSelector, createSlice, PayloadAction } from '@reduxjs/toolkit'
import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux'
import { AdjustmentKey, AdjustmentsState, applyAdjustments } from './adjustments'
import { PersistConfig, persistReducer, PersistedState } from 'redux-persist'
import { createIDBStorage } from '@/common/rp-storage-idb'

const NAME = 'icePhotoEditor' as const

interface AdjustmentChange {
  name: AdjustmentKey
  prevValue: Exclude<AdjustmentsState[AdjustmentKey], undefined>
  currentValue: Exclude<AdjustmentsState[AdjustmentKey], undefined>
}

interface HistoryElement {
  date: number
  adjustment: AdjustmentChange
}

interface IcePhotoEditorState {
  /**
   * The original image that is never modified
   */
  image?: ImageData
  /**
   * The image that is the result of all adjustments applied
   */
  currentImage?: ImageData
  historyStack: HistoryElement[]
  currentHistoryPointer: number
  loadedDate?: number
  filename?: string
}

// This slice will be injected asyncronously which is why we need separate types here
interface SliceState {
  [NAME]: IcePhotoEditorState & Partial<PersistedState>
}
export type CurrentRootState = MakeCurrentState<SliceState>
export type CurrentDispatch = MakeCurrentDispatch<SliceState>
// eslint-disable-next-line @typescript-eslint/no-invalid-void-type
export type CurrentThunk<ReturnType = void> = MakeCurrentThunk<SliceState, ReturnType>

export const useCurrentDispatch: () => CurrentDispatch = useDispatch
export const useCurrentSelector: TypedUseSelectorHook<CurrentRootState> = useSelector

const initialState: IcePhotoEditorState = {
  historyStack: [],
  currentHistoryPointer: -1,
}

const persistConfig: PersistConfig<IcePhotoEditorState> = {
  key: NAME,
  // localStorage doesn't have enough space for this store
  storage: createIDBStorage(NAME),
  serialize: false,
  deserialize: false,
} as const

/*

REDUCERS

*/

const { actions, reducer } = createSlice({
  name: NAME,
  initialState,
  reducers: {
    loadImage: (state, action: PayloadAction<{ imageData: ImageData; filename: string }>) => {
      Object.assign(state, initialState)
      state.currentImage = state.image = action.payload.imageData
      state.filename = action.payload.filename

      state.loadedDate = Date.now()
    },
    addHistoryElement: (state, action: PayloadAction<HistoryElement>) => {
      state.historyStack.push(action.payload)
      state.currentHistoryPointer = state.historyStack.length - 1
    },
    updateLastHistoryElement: (state, action: PayloadAction<HistoryElement>) => {
      const { adjustment: lastAdjustment } = state.historyStack.at(-1) ?? {}

      if (!lastAdjustment) return

      Object.assign(lastAdjustment, action.payload.adjustment)
    },
    setCurrentHistoryElement: (state, action: PayloadAction<number>) => {
      const historyElementIndex = state.historyStack.findIndex(
        ({ date }) => date === action.payload
      )

      state.currentHistoryPointer = historyElementIndex

      if (action.payload === state.loadedDate) {
        state.currentHistoryPointer = -1
      }
    },
    applyAdjustments: (state) => {
      if (!state.image) return

      const historyStack = state.historyStack.slice(0, state.currentHistoryPointer + 1)

      const adjustmentsState = historyStack.reduce<AdjustmentsState>(
        (result, { adjustment }) => ({
          ...result,
          [adjustment.name]: adjustment.currentValue,
        }),
        {}
      )

      // eslint-disable-next-line unicorn/prefer-spread
      const newImageData = applyAdjustments(state.image.data.slice(0), adjustmentsState)

      state.currentImage = new ImageData(newImageData, state.image.width, state.image.height, {
        colorSpace: state.image.colorSpace,
      })
    },
  },
})

export { actions }

store.injectReducer(NAME, persistReducer(persistConfig, reducer))

/*

THUNKS

*/

export const updateAdjustments =
  (adjustment: AdjustmentChange): CurrentThunk =>
  (dispatch, getState) => {
    const { historyStack } = getState()[NAME]

    const { adjustment: lastAdjustment } = historyStack.at(-1) ?? {}
    const { adjustment: prevAdjustment } =
      historyStack.findLast(
        (historyElement) =>
          historyElement.adjustment.name === adjustment.name &&
          historyElement.adjustment !== lastAdjustment
      ) ?? {}

    if (prevAdjustment) {
      adjustment.prevValue = prevAdjustment.currentValue
    }

    if (lastAdjustment && lastAdjustment.name === adjustment.name) {
      dispatch(actions.updateLastHistoryElement({ date: Date.now(), adjustment }))
    } else {
      dispatch(actions.addHistoryElement({ date: Date.now(), adjustment }))
    }

    dispatch(actions.applyAdjustments())
  }

/*

SELECTORS

*/

export const selectImage = (state: CurrentRootState) => state[NAME].image

export const selectCurrentImage = (state: CurrentRootState) => state[NAME].currentImage

export const selectHistoryStack = (state: CurrentRootState) => state[NAME].historyStack

export const selectLoadedDate = (state: CurrentRootState) => state[NAME].loadedDate

export const selectFilename = (state: CurrentRootState) => state[NAME].filename

export const selectPersistRehydrated = (state: CurrentRootState) => state[NAME]._persist?.rehydrated

export const selectCurrentHistoryPointer = (state: CurrentRootState) =>
  state[NAME].currentHistoryPointer

export const selectHistory = createSelector(
  selectHistoryStack,
  selectLoadedDate,
  (historyStack, loadedDate) => [
    ...(loadedDate
      ? [
          {
            date: loadedDate,
            adjustment: {
              name: 'Image Loaded',
              prevValue: '',
              currentValue: '',
            },
          },
        ]
      : []),
    ...historyStack,
  ]
)

export const selectAdjustmentsStateStrings = createSelector(
  selectHistoryStack,
  selectCurrentHistoryPointer,
  (historyStack, currentHistoryPointer) =>
    historyStack.slice(0, currentHistoryPointer + 1).reduce<Partial<Record<AdjustmentKey, string>>>(
      (result, { adjustment }) => ({
        ...result,
        [adjustment.name]: adjustment.currentValue.toString(),
      }),
      {}
    )
)
