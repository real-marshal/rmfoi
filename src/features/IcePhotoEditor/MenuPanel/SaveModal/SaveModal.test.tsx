import store from '@/app/store'
import { composeStories } from '@storybook/react'
import { render } from '@testing-library/react'
import { Provider } from 'react-redux'
import * as stories from './SaveModal.stories'

describe('SaveModal component', () => {
  const testStories = Object.values(composeStories(stories)).map(
    (Story) => [Story.storyName, Story] as [string, typeof Story]
  )

  it.each(testStories)('renders %s story', (_, Story) => {
    const view = render(
      <Provider store={store}>
        <Story />
      </Provider>
    )
    expect(view.baseElement).toMatchSnapshot()
  })
})
