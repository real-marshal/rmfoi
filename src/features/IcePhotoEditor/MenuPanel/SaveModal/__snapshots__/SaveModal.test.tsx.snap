// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SaveModal component renders Basic story 1`] = `
<body>
  <div />
</body>
`;
